node/overseer/overseer-gen/proc-macro/src/impl_builder.rs

```diff
@@ -107,46 +76,235 @@ pub(crate) fn impl_builder(info: &OverseerInfo) -> proc_macro2::TokenStream {
 		})
 		.collect::<Vec<_>>();
 
-	let generics = quote! {
-		< S, #( #baggage_generic_ty, )* >
-	};
-	let where_clause = quote! {
+	// Helpers to use within quote! macros
+	let builder_where_clause = quote! {
 		where
-			S: #support_crate ::SpawnNamed,
+			#(
+				#subsystem_type : Subsystem<#subsystem_ctx_name< #consumes >, #error_ty>,
+			)*
+			S: #support_crate ::SpawnNamed + Send
 	};
-
 	let builder_generics = quote! {
-		<S, #( #baggage_generic_ty, )* #( #builder_generic_ty, )* >
+		S, #( #subsystem_type, )* #( #baggage_generic_ty, )*
 	};
 
-	// all subsystems must have the same context
-	// even if the overseer does not impose such a limit.
-	let builder_additional_generics = quote! {
-		<#( #builder_generic_ty, )* >
-	};
+	let field_name = subsystem_name.iter().chain(baggage_name.iter()).collect::<Vec<_>>();
+	let field_type = subsystem_type
+		.iter()
+		.chain(info.baggage().iter().filter_map(|bag| bag.field_ty.get_ident()))
+		.collect::<Vec<_>>();
 
-	let consumes = &info.consumes();
+	// Setters logic
 
-	let subsyste_ctx_name =
-		Ident::new(&(overseer_name.to_string() + "SubsystemContext"), overseer_name.span());
+	// For each setter we need to leave the remaining fields untouched and
+	// remove the field that we are fixing in this setter
+	// For subsystems we also need _with and replace_ setters
+	let subsystem_specific_setters = info
+		.subsystems()
+		.iter()
+		.filter(|ssf| !ssf.wip)
+		.enumerate()
+		.map(|(idx, ssf)| {
+			let fname = &ssf.name;
+			let ftype = &ssf.generic;
+			let subsystem_consumes = &ssf.consumes;
+			// Remove placeholder at specific field position
+			let impl_generics = &(subsystem_placeholder_ty[..idx]
+				.iter()
+				.chain(subsystem_placeholder_ty[idx + 1..].iter())
+				.collect::<Vec<_>>());
```

Comment:
nit: I do not like this. It looks / feels very unidiomatic. It also changes the order of generics. `let mut subsystem_placeholder_ty = subsystem_placeholder_ty.clone(); subsystem_placeholder_ty[idx] = foo;` reads significantly better.

Reply:
This takes all placeholders but one. It could be done via `clone+remove`, of course.

Reply:
Why does it take all but one? Why not all and replace the one in question?